### PR TITLE
headers: consistent validation

### PR DESCRIPTION
**Pull request by alyssawilk**
Risk Level: low
Testing: existing tests
Docs Changes: n/a
Release Notes: inline
[Optional Runtime guard:] yes

### CODE REVIEW

File: source/extensions/filters/common/expr/context.h [hunk 1]
```diff
@@ -116,9 +117,16 @@ template <class T> class HeadersWrapper : public google::api::expr::runtime::Cel
       return {};
     }
     auto str = std::string(key.StringOrDie().value());
-    if (!::Envoy::Http::validHeaderString(str)) {
-      // Reject key if it is an invalid header string
-      return {};
+    if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.consistent_header_validation")) {
+      if (!Http::HeaderUtility::headerValueIsValid(str)) {
```
Comment (tyxia) [inline source/extensions/filters/common/expr/context.h:121]: The intention here is checking the **header key/name** . So I think we should use utility `headerNameIsValid` instead?

File: test/extensions/filters/common/expr/context_test.cc [hunk 2]
```diff
@@ -50,6 +51,17 @@ TEST(Context, InvalidRequest) {
...
 }
 
+TEST(Context, InvalidRequestLegacy) {
+  TestScopedRuntime scoped_runtime;
+  scoped_runtime.mergeValues({{"envoy.reloadable_features.consistent_header_validation", "false"}});
+
+  Http::TestRequestHeaderMapImpl header_map{{"referer", "dogs.com"}};
+  Protobuf::Arena arena;
+  HeadersWrapper<Http::RequestHeaderMap> headers(arena, &header_map);
+  auto header = headers[CelValue::CreateStringView("dogs.com\n")];
```
Comment (tyxia) [inline test/extensions/filters/common/expr/context_test.cc:61]: We probably should test a invalid header key/name in particular.

**Review by tyxia — COMMENTED** [review]
Thanks for improving code health!

Some drive-by comments

File: source/extensions/filters/common/expr/context.h (hunk 1 above)
Comment (alyssawilk) [reply source/extensions/filters/common/expr/context.h]: yeah good catch - I'd been going off the bug which I think referenced the wrong helper.  looks like same thing with the formatter.

File: test/extensions/filters/common/expr/context_test.cc (hunk 2 above)
Comment (alyssawilk) [reply test/extensions/filters/common/expr/context_test.cc]: so this does exercise the test code.  did you just prefer an invalid string that looks like a key? PTAL

File: source/extensions/filters/common/expr/context.h (hunk 1 above)
Comment (tyxia) [reply source/extensions/filters/common/expr/context.h]: Yeah,  after tracking the code into [here](https://github.com/envoyproxy/envoy/blob/e69057a11361359e55f9cd4ba96bf70be1ea5e76/source/common/formatter/http_specific_formatter.cc#L85-L88) , we can see formatter is also using header key/name

File: test/extensions/filters/common/expr/context_test.cc (hunk 2 above)
Comment (tyxia) [reply test/extensions/filters/common/expr/context_test.cc]: Yeah, i was preferring a invalid string that looks like a key. Thanks for updates.

**Review by tyxia — APPROVED** [review]
Thanks!

Defer to @yanavlasov for final review and merge.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
LGTM with a spelling fix

**Comment by yanavlasov** [comment]
Needs fixing merge conflict.

/wait

**Pull request merged**